### DISCUSSION TITLE
UniMath: general or domain-specific?

### DISCUSSION THREAD

**Discussion by benediktahrens** [UniMath/UniMath#1628] [General]
I am currently preparing a talk for MAP at IPAM, UCLA. For this, I looked at UniMath at bit, and found out that 94.6 % of code (counting only definitions and proofs) is in the packages `CategoryTheory` and `Bicategories`. I thought I'd share this, as I found it somewhat surprising.

**Comment by nmvdw** [comment]
Do you have number of lines of code for each separate package? Do you include the packages Foundations and MoreFoundations in this percentage?

I think UniMath is a general purpose library. However, the people involved just chose to mainly work on (bi)category theory. Contributions on all fields of mathematics are welcome.

**Reply by benediktahrens** [reply]
No, I don't have per-package numbers, just the total and the ones for (CategoryTheory + Bicategories). The total includes all the packages and comes from `$ make coqwc`.

I agree that all fields of mathematics are welcome.

**Reply by nmvdw** [reply]
My count
```
Package                  Definition          Proof            Total      Percentage
-----------------------------------------------------------------------------------
Algebra                        7858          14060            21918            6.4%
AlgebraicGeometry               244            341              585           0.17%
Bicategories                  49151          90610           139761           40.5%
CategoryTheory                41123          66046           107169           31.1%
Combinatorics                  2604           5566             8170            2.4%
Folds                           334            651              985           0.29%
Foundations                    3615           6790            10405              3%
HomologicalAlgebra             4167           9022            13189            3.8%
Induction                       701           1522             2223           0.64%
Ktheory                          51             55              106           0.03%
MoreFoundations                2125           3092             5217            1.5%
NumberSystems                   743           1008             1751           0.51%
PAdics                          890           6849             7739            2.2%
Paradoxes                        65             83              148          0.043%
RealNumbers                    2082           7279             9361            2.7%
SubstitutionSystems            4206           6383            10589            3.1%
SyntheticHomotopyTheory         586            710             1296           0.38%
Tactics                        1168              9             1175           0.34%
Topology                        862           2018             2880           0.84%
-----------------------------------------------------------------------------------
Total                                                        344667
```

**Reply by benediktahrens** [reply]
Interesting, these numbers look very different from the ones I obtained, and much more realistic. 

In detail, I did the following
1. I called `$ make coqwc` and added the numbers for definitions and proofs.
2. I called `$ $ coqwc UniMath/CategoryTheory/*.v UniMath/CategoryTheory/*/*.v UniMath/CategoryTheory/*/*/*.v UniMath/Bicategories/*.v UniMath/Bicategories/*/*.v UniMath/Bicategories/*/*/*.v UniMath/Bicategories/*/*/*.v` and added the numbers for definitions and proofs.

That gave me 345457 and 326865, respectively. I wonder why these numbers are so weird...

**Reply by nmvdw** [reply]
Could it be because you counted `UniMath/Bicategories/*/*/*.v` twice?

**Reply by benediktahrens** [reply]
That's it, thanks a lot! That would have been an embarrassing mistake.

Did you write a script to create that table? If so, could we add it to the repository?

**Reply by nmvdw** [reply]
I did not create a script. It was mostly by hand.
